design/src/rfcs/rfc0026_client_crate_organization.md

```diff
@@ -0,0 +1,396 @@
+    |   └── <One struct per error named `${error}`>
+    ├── builders
+    |   └── <One struct per shape named `${shape}Builder`>
+    └── <One struct per shape>
+```
+
+Customers using the fluent builder should be able to just `use ${crate}::types::*;` to immediately
+get access to all the shared types needed by the operations they are calling.
+
+Additionally, moving the top-level code generated error types into `crate::types` will eliminate a name
+collision issue in the `crate::error` module.
+
+### Repurpose the original `crate::error` module
+
+The `error` module is significantly smaller after all the code generated error types
+are moved out of it. This top-level module is now available for re-exports and utilities.
+
+The following will be re-exported in `crate::error`:
+- `aws_smithy_http::result::SdkError`
+- `aws_smithy_types::error::display::DisplayErrorContext`
+
+For crates that have an `ErrorExt`, it will also be moved into `crate::error`.
+
+### Flatten the `presigning` module
+
+The `crate::presigning` module only has four members, so it should be flattened from:
+
+```text
+.
+└── presigning
+    ├── config
+    |   ├── Builder
+    |   ├── Error
+    |   └── PresigningConfig
+    └── request
+        └── PresignedRequest
+```
+
+to:
+
+```text
+.
+└── presigning
+    ├── PresigningConfigBuilder
+    ├── PresigningConfigError
+    ├── PresigningConfig
+    └── PresignedRequest
+```
+
+At the same time, `Builder` and `Error` will be renamed to `PresigningConfigBuilder` and `PresigningConfigError`
+respectively since these will rarely be referred to directly (preferring `PresigningConfig::builder()` instead; the
+error will almost always be unwrapped).
+
+### Remove the empty modules
+
+The `lens` and `http_body_checksum` modules have nothing inside them,
+and their documentation descriptions are not useful to customers:
+
+> `lens`: Generated accessors for nested fields
+
+> `http_body_checksum`: Functions for modifying requests and responses for the purposes of checksum validation
+
+These modules hold private functions that are used by other generated code, and should just be made
+private or `#[doc(hidden)]` if necessary.
+
+New Organization
+----------------
+
+All combined, the following is the new publicly visible organization:
+
+```text
+.
+├── client
+|   ├── Builder (only in non-SDK crates) (*)
+|   └── Client
+├── config
+|   ├── retry
+|   |   ├── RetryConfig (*)
+|   |   ├── RetryConfigBuilder (*)
+|   |   └── RetryMode (*)
+|   ├── timeout
+|   |   ├── TimeoutConfig (*)
+|   |   └── TimeoutConfigBuilder (*)
+|   ├── AppName (*)
+|   ├── AsyncSleep (*)
+|   ├── Builder
+|   ├── Config
+|   ├── Credentials (*)
+|   ├── Endpoint (*)
+|   ├── Region (*)
+|   └── Sleep (*)
+├── error
+|   ├── DisplayErrorContext (*)
+|   ├── ErrorExt (for some services)
+|   └── SdkError (*)
+├── meta
+|   └── PKG_VERSION
+├── middleware
+|   └── DefaultMiddleware
+├── operation
+|   ├── customize
```

Comment:
From an offline discussion: `customize` should be moved into the `client` module to better avoid name collisions.